### PR TITLE
Sdi 499 create activity api

### PR DESCRIPTION
**Pull request by steverendell**
New activity sync service

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/data/CreateActivityRequest.kt [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+  @Schema(description = "Activity end date", example = "2022-08-12")
+  val endDate: LocalDate?,
+
+  @Schema(description = "Prison where the activity is to occur", required = true)
+  @NotBlank
+  val prisonId: String,
+
+  @Schema(description = "Room where the activity is to occur (from activity schedule)", required = true)
+  @NotNull
+  val internalLocation: Long,
```
Comment (andymarke) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/data/CreateActivityRequest.kt:31]: interesting, will "they" be providing this or will it be mapped from their own room description?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/jpa/CourseActivity.kt [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+  @Column(name = "CRS_ACTY_ID", nullable = false)
+  val courseActivityId: Long = 0,
+
+  @Column
+  @Size(max = 20)
+  var code: String? = null,
+
+  @Column
+  @Size(max = 6)
+  var caseloadId: String? = null,
```
Comment (andymarke) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/jpa/CourseActivity.kt:66]: do you know what this is used for?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/jpa/CourseActivity.kt [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+  val scheduleStartDate: LocalDate? = null,
+
+  @Column
+  val scheduleEndDate: LocalDate? = null,
+
+  @OneToMany(mappedBy = "courseActivity", cascade = [CascadeType.ALL], fetch = FetchType.LAZY)
+  var payRates: List<CourseActivityPayRate>? = null,
+
+  @Column
+  @Size(max = 12)
```
Comment (andymarke) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/jpa/CourseActivity.kt:97]: as a matter of interest do the `size` annotations have any impact on anything?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/data/CreateActivityRequest.kt [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+
+  @Schema(description = "Pay rates", required = true)
+  val payRates: List<PayRateRequest>,
+
+  @Schema(description = "Description from concatenated activity and activity schedule", required = true)
+  @NotBlank
+  val description: String,
+
+  @Schema(description = "Minimum Incentive Level")
+  val minimumIncentiveLevel: String? = null,
```
Comment (andymarke) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/data/CreateActivityRequest.kt:44]: maybe minimumIncentiveLevelCode?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/jpa/CourseActivity.kt [hunk 5]
```diff
@@ -0,0 +1,175 @@
...
+    other as CourseActivity
+
+    return courseActivityId == other.courseActivityId
+  }
+
+  override fun hashCode(): Int = javaClass.hashCode()
+
+  @Override
+  override fun toString(): String {
+    return this::class.simpleName + "(courseActivityId = $courseActivityId )"
```
Comment (andymarke) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/jpa/CourseActivity.kt:153]: maybe add description? might be useful when debugging?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/jpa/CourseActivityPayRate.kt [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+
+  @Id
+  @ManyToOne(optional = false, fetch = FetchType.LAZY)
+  @JoinColumn(name = "CRS_ACTY_ID", nullable = false)
+  var courseActivity: CourseActivity,
+
+  @Id
+  @Column(name = "IEP_LEVEL")
+  @Size(max = 12)
+  val iepLevelCode: String,
```
Comment (andymarke) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/jpa/CourseActivityPayRate.kt:28]: do you get different pay rates per IEP level?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/jpa/CourseActivity.kt [hunk 7]
```diff
@@ -0,0 +1,175 @@
...
+  @Column
+  var capacity: Int? = null,
+
+  @Column(name = "ACTIVE_FLAG")
+  @Type(type = "yes_no")
+  val active: Boolean = false,
+
+  @ManyToOne(optional = false)
+  @JoinColumn(name = "PROGRAM_ID", nullable = false)
+  var program: ProgramService? = null,
```
Comment (lbennett-moj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/jpa/CourseActivity.kt:85]: val as not nullable?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/jpa/CourseActivity.kt [hunk 8]
```diff
@@ -0,0 +1,175 @@
...
+        )
+      ),
+      JoinColumnOrFormula(column = JoinColumn(name = "IEP_LEVEL", referencedColumnName = "code"))
+    ]
+  )
+  val iepLevel: IEPLevel,
+
+  @ManyToOne(optional = false)
+  @JoinColumn(name = "INTERNAL_LOCATION_ID", nullable = false)
+  var internalLocation: AgencyInternalLocation? = null,
```
Comment (lbennett-moj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/jpa/CourseActivity.kt:132]: Would this ever change - maybe val and not nullable

File: src/main/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/data/CreateActivityRequest.kt [hunk 9]
```diff
@@ -0,0 +1,118 @@
...
+  @Schema(description = "Program Service code (from activity category)", required = true)
+  @NotBlank
+  val programCode: String,
+)
+
+@JsonInclude(JsonInclude.Include.NON_NULL)
+@Schema(description = "Course activity creation request pay rates")
+data class PayRateRequest(
+  @Schema(description = "The incentive level", example = "2022-08-12")
+  val incentiveLevel: String? = null,
```
Comment (lbennett-moj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/data/CreateActivityRequest.kt:55]: example text typo

File: src/main/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/data/CreateActivityRequest.kt (hunk 1 above)
Comment (lbennett-moj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/data/CreateActivityRequest.kt:31]: is this id?  maybe internalLocationId

File: src/main/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/service/ActivitiesService.kt [hunk 10]
```diff
@@ -0,0 +1,103 @@
...
+      null
+    )
+    log.debug("Activity created with Nomis id = (${courseActivity.courseActivityId})")
+
+    return CreateActivityResponse(courseActivity.courseActivityId)
+  }
+
+  private fun mapModel(dto: CreateActivityRequest): CourseActivity {
+
+    val prison = agencyLocationRepository.findById(dto.prisonId)
```
Comment (andymarke) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/service/ActivitiesService.kt:55]: we tend to use `findByIdOrNull` to use kotlin idiomatic style

File: src/test/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/resource/ActivitiesResourceIntTest.kt [hunk 11]
```diff
@@ -0,0 +1,157 @@
...
+import uk.gov.justice.digital.hmpps.nomisprisonerapi.data.CreateActivityResponse
+import uk.gov.justice.digital.hmpps.nomisprisonerapi.data.PayRateRequest
+import uk.gov.justice.digital.hmpps.nomisprisonerapi.helper.builders.Repository
+import uk.gov.justice.digital.hmpps.nomisprisonerapi.integration.IntegrationTestBase
+import uk.gov.justice.digital.hmpps.nomisprisonerapi.jpa.ProgramService
+import java.math.BigDecimal
+import java.time.LocalDate
+
+private const val prisonId = "LEI"
+private const val roomId: Long = -8
```
Comment (andymarke) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/resource/ActivitiesResourceIntTest.kt:23]: where does this -8 come from?

**Review by andymarke — APPROVED** [review]
Looks good.

I think generally for the JPA can we keep everything to `val` until we need to mutate them? Also looks like you have a few nullable elements in JPA data classes that look like they are non-null columns so shouldn't they not be nullable?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/jpa/CourseActivity.kt (hunk 7 above)
Comment (steverendell) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/jpa/CourseActivity.kt:85]: 👍

File: src/main/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/data/CreateActivityRequest.kt (hunk 1 above)
Comment (steverendell) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/data/CreateActivityRequest.kt:31]: they are providing it !

File: src/main/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/jpa/CourseActivity.kt (hunk 2 above)
Comment (steverendell) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/jpa/CourseActivity.kt:66]: No but looks like it needs to be populated

File: src/main/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/data/CreateActivityRequest.kt (hunk 1 above)
Comment (steverendell) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/data/CreateActivityRequest.kt:31]: good point

File: src/main/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/jpa/CourseActivity.kt (hunk 3 above)
Comment (steverendell) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/jpa/CourseActivity.kt:97]: They just seemed a good idea at the time but happy to remove them

File: src/main/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/data/CreateActivityRequest.kt (hunk 4 above)
Comment (steverendell) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/data/CreateActivityRequest.kt:44]: 👍

File: src/main/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/jpa/CourseActivity.kt (hunk 5 above)
Comment (steverendell) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/jpa/CourseActivity.kt:153]: 👍

File: src/main/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/jpa/CourseActivityPayRate.kt (hunk 6 above)
Comment (steverendell) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/jpa/CourseActivityPayRate.kt:28]: Dont know, but iep is part of the composite primary key so seems to be important

File: src/test/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/resource/ActivitiesResourceIntTest.kt (hunk 11 above)
Comment (steverendell) [reply src/test/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/resource/ActivitiesResourceIntTest.kt:23]: seed data - added comment.

**Review by andymarke — APPROVED** [review]
LGTM

**Review by andymarke — APPROVED** [review]
LGTM

**Pull request merged**